feat: 캐싱 모듈 구현

psangwon62 opened the pull request:
## What is this PR?
- ASCacheKit을 구현하였습니다. #12 

## PR Type
- [ ] Bugfix
- [ ] Chore
- [x] New feature (기능을 추가하는 feat)
- [ ] Breaking change (기존의 기능이 동작하지 않을 수 있는 fix/feat)
- [ ] Documentation Update

## ScreenShot(if available)
|테스트|스크린샷|
|:--:|:--:|
|GIF|<img src = "https://github.com/user-attachments/assets/936a8f35-1726-4cf2-a6a3-9ebedb65d5d4" width ="250">|

## Further comments
- 모듈 사용 문서를 작성할 계획입니다.
- 테스트해 볼 수 있는 데모 앱을 추가할 예정입니다.

psangwon62 committed:
feat: 캐싱 모듈 프로젝트에 추가

psangwon62 committed:
feat: final class로 변경

psangwon62 committed:
feat: 테스트를 위한 Mock 객체 생성

psangwon62 committed:
feat: ASCacheManager test code 작성

- 테스트 코드를 작성했습니다.
- CacheManager -> ASCacheManager로 이름을 변경했습니다.
- 테스트를 위한 ASCacheManager 이니셜라이저를 생성하였습니다.
- URLSessin Mock 객체를 위한 URLSessionProtocol을 작성하고 적용하였습니다.
- 캐싱 시 사용하는 키를 SHA256으로 암호화하였습니다.

psangwon62 committed:
chore: 사용하지 않는 import문 제거

Sonny-Kor commented on ASCacheKit/Core/Source/String+Extension.swift:
p3(웬만하면 반영해 주세요) ;
파일명에 String+ "어떤 Extension" 인지 구분해서 쓰면 좋은것 같은데 어떠신가용

psangwon62 replied:
Extension 폴더를 만들고, String+Encrypt로 변경했습니다!

Tltlbo commented on ASCacheKit/Core/Source/ASCacheManager.swift:
```diff
@@ -0,0 +1,83 @@
+        self.urlSession = session
+    }
+
+    public func loadImage(from url: URL, cacheOption: CacheOption) async -> UIImage? {
+        let cacheKey = url.absoluteString
+        if let data = await loadData(forKey: cacheKey, cacheOption: cacheOption) {
+            return UIImage(data: data)
+        }
+        return nil
+    }
```
P2: 다른 부분은 Data를 리턴하고 있는 듯 한데 얘만 UIImage를 리턴하고 있는 것 같습니다.
이 친구도 Data를 리턴해서 Data를 받는 측에서 UIImage에 Data를 넣는 방식이 어떨까요?

psangwon62 replied:
좋은 생각인 것 같아요. loadImage, loadAudio 함수를 loadCache로 통합해 보겠습니다

Tltlbo commented on ASCacheKit/Core/Source/DiskCache.swift:
```diff
@@ -0,0 +1,36 @@
+import UIKit
```
P1: UIKit이 필요한 부분이 있을까요?

psangwon62 replied:
UIImage를 저장해보려고 하다가 남겨놨나 봅니다,, Foundation으로 변경완료!

Tltlbo commented on ASCacheKit/Core/Source/MemoryCache.swift:
```diff
@@ -0,0 +1,17 @@
+import UIKit
```
P1: 위의 코멘트와 동일!

psangwon62 replied:
Foundation으로 변경완료!

Tltlbo reviewed (commented):
고생하셨습니다!

psangwon62 committed:
refactor: 코드 리뷰 반영

- 필요없는 UIKit import를 Foundation으로 변경
- loadImage, loadAudio를 loadCache로 통합
- 테스트 함수 변수명 변경
- Extension 폴더 생성 및 String Extension 파일명 변경
